pkg/logtailing/tailer.go

```diff
@@ -107,60 +105,84 @@ func New(cfg *Config) *Tailer {
+	s := ansi.StartSpinner("Getting ready...", t.cfg.Log.Out)
+
+	ctx = withSIGTERMCancel(ctx, func() {
 		log.WithFields(log.Fields{
 			"prefix": "logtailing.Tailer.Run",
 		}).Debug("Ctrl+C received, cleaning up...")
+	})
 
-		<-t.interruptCh
-		cancel()
-	}()
+	var warned = false
+	var nAttempts int = 0
 
-	// Create the CLI session
-	session, err := t.createSession()
-	if err != nil {
-		ansi.StopSpinner(s, "", t.cfg.Log.Out)
-		t.cfg.Log.Fatalf("Error while authenticating with Stripe: %v", err)
-	}
+	for nAttempts < maxConnectAttempts {
+		session, err := t.createSession(ctx)
 
-	// Create and start the websocket client
-	t.webSocketClient = websocket.NewClient(
-		session.WebSocketURL,
-		session.WebSocketID,
-		session.WebSocketAuthorizedFeature,
-		&websocket.Config{
-			EventHandler:      websocket.EventHandlerFunc(t.processRequestLogEvent),
-			Log:               t.cfg.Log,
-			NoWSS:             t.cfg.NoWSS,
-			ReconnectInterval: time.Duration(session.ReconnectDelay) * time.Second,
-		},
-	)
-	go t.webSocketClient.Run()
-
-	select {
-	case <-t.webSocketClient.Connected():
-		ansi.StopSpinner(s, "Ready! You're now waiting to receive API request logs (^C to quit)", t.cfg.Log.Out)
-	case <-t.ctx.Done():
-		ansi.StopSpinner(s, "", t.cfg.Log.Out)
-		t.cfg.Log.Fatalf("Aborting")
-	}
+		if err != nil {
+			ansi.StopSpinner(s, "", t.cfg.Log.Out)
+			t.cfg.Log.Fatalf("Error while authenticating with Stripe: %v", err)
+		}
 
-	if session.DisplayConnectFilterWarning {
-		color := ansi.Color(os.Stdout)
-		fmt.Printf("%s you specified the 'account' filter for Connect accounts but are not a Connect user, so the filter will not be applied.\n", color.Yellow("Warning"))
-	}
+		if session.DisplayConnectFilterWarning && !warned {
+			color := ansi.Color(os.Stdout)
+			fmt.Printf("%s you specified the 'account' filter for Connect accounts but are not a Connect user, so the filter will not be applied.\n", color.Yellow("Warning"))
+			// Only display this warning once
+			warned = true
+		}
 
-	// Block until context is done (i.e. Ctrl+C is pressed)
-	<-t.ctx.Done()
+		t.webSocketClient = websocket.NewClient(
+			session.WebSocketURL,
+			session.WebSocketID,
+			session.WebSocketAuthorizedFeature,
+			&websocket.Config{
+				EventHandler:      websocket.EventHandlerFunc(t.processRequestLogEvent),
+				Log:               t.cfg.Log,
+				NoWSS:             t.cfg.NoWSS,
+				ReconnectInterval: time.Duration(session.ReconnectDelay) * time.Second,
+			},
+		)
+
+		go func() {
+			<-t.webSocketClient.Connected()
+			nAttempts = 0
+			ansi.StopSpinner(s, "Ready! You're now waiting to receive API request logs (^C to quit)", t.cfg.Log.Out)
+		}()
+
+		go t.webSocketClient.Run(ctx)
+		nAttempts++
+
+		select {
+		case <-ctx.Done():
+			ansi.StopSpinner(s, "", t.cfg.Log.Out)
+			t.cfg.Log.Fatalf("Aborting")
+		case <-t.webSocketClient.NotifyExpired:
+			if nAttempts < maxConnectAttempts {
+				s = ansi.StartSpinner("Session expired, reconnecting...", t.cfg.Log.Out)
```

Comment:
Do we need to stop the old spinner first? Or does starting a new spinner automatically stops the existing one?

Reply:
Good call. I had not been testing with color on. It was a jittery mess. I fixed it in the latest.